<h1>Содержание<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Загрузка-данных" data-toc-modified-id="Загрузка-данных-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Загрузка данных</a></span></li><li><span><a href="#Умножение-матриц" data-toc-modified-id="Умножение-матриц-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Умножение матриц</a></span></li><li><span><a href="#Алгоритм-преобразования" data-toc-modified-id="Алгоритм-преобразования-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Алгоритм преобразования</a></span></li><li><span><a href="#Проверка-алгоритма" data-toc-modified-id="Проверка-алгоритма-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Проверка алгоритма</a></span></li><li><span><a href="#Финальный-вывод" data-toc-modified-id="Финальный-вывод-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>Финальный вывод</a></span></li><li><span><a href="#Чек-лист-проверки" data-toc-modified-id="Чек-лист-проверки-6"><span class="toc-item-num">6&nbsp;&nbsp;</span>Чек-лист проверки</a></span></li></ul></div>

# Защита персональных данных клиентов

Вам нужно защитить данные клиентов страховой компании «Хоть потоп». Разработайте такой метод преобразования данных, чтобы по ним было сложно восстановить персональную информацию. Обоснуйте корректность его работы.

Нужно защитить данные, чтобы при преобразовании качество моделей машинного обучения не ухудшилось. Подбирать наилучшую модель не требуется.

Признаки: пол, возраст и зарплата застрахованного, а также количество членов его семьи.

Целевой признак: количество страховых выплат клиенту за последние 5 лет.

Соответственно данные представлены в виде пяти столбцов, четыре из которых - признаки, а один оставшийся - целевой признак.

## Загрузка данных

In [1]:
import pandas as pd

#import seaborn as sns

import numpy as np

import matplotlib.pyplot as plt

from sklearn.linear_model import LinearRegression
from sklearn.metrics import r2_score

In [2]:
data = pd.read_csv('/datasets/insurance.csv')

In [3]:
data.head(20)

Пол  Возраст  Зарплата  Члены семьи  Страховые выплаты
0     1     41.0   49600.0            1                  0
1     0     46.0   38000.0            1                  1
2     0     29.0   21000.0            0                  0
3     0     21.0   41700.0            2                  0
4     1     28.0   26100.0            0                  0
5     1     43.0   41000.0            2                  1
6     1     39.0   39700.0            2                  0
7     1     25.0   38600.0            4                  0
8     1     36.0   49700.0            1                  0
9     1     32.0   51700.0            1                  0
10    1     25.0   36600.0            1                  0
11    1     38.0   29300.0            0                  0
12    0     23.0   39500.0            3                  0
13    0     21.0   55000.0            0                  0
14    0     40.0   43700.0            1                  0
15    1     34.0   23300.0            0                  0
16    1     26.0   48900.0            2                  0
17    1     41.0   33200.0            2                  0
18    1     42.0   49700.0            0                  1
19    1     27.0   36900.0            0                  0

In [4]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5000 entries, 0 to 4999
Data columns (total 5 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Пол                5000 non-null   int64  
 1   Возраст            5000 non-null   float64
 2   Зарплата           5000 non-null   float64
 3   Члены семьи        5000 non-null   int64  
 4   Страховые выплаты  5000 non-null   int64  
dtypes: float64(2), int64(3)
memory usage: 195.4 KB


In [5]:
data.describe()

Пол      Возраст      Зарплата  Члены семьи  Страховые выплаты
count  5000.000000  5000.000000   5000.000000  5000.000000        5000.000000
mean      0.499000    30.952800  39916.360000     1.194200           0.148000
std       0.500049     8.440807   9900.083569     1.091387           0.463183
min       0.000000    18.000000   5300.000000     0.000000           0.000000
25%       0.000000    24.000000  33300.000000     0.000000           0.000000
50%       0.000000    30.000000  40200.000000     1.000000           0.000000
75%       1.000000    37.000000  46600.000000     2.000000           0.000000
max       1.000000    65.000000  79000.000000     6.000000           5.000000

</font><font color='Black'><b>Пояснение</b></font><br>
<font color='Black'>Поищем дубликаты

In [6]:
data[data.duplicated()]

Пол  Возраст  Зарплата  Члены семьи  Страховые выплаты
281     1     39.0   48100.0            1                  0
488     1     24.0   32900.0            1                  0
513     0     31.0   37400.0            2                  0
718     1     22.0   32600.0            1                  0
785     0     20.0   35800.0            0                  0
...   ...      ...       ...          ...                ...
4793    1     24.0   37800.0            0                  0
4902    1     35.0   38700.0            1                  0
4935    1     19.0   32700.0            0                  0
4945    1     21.0   45800.0            0                  0
4965    0     22.0   40100.0            1                  0

[153 rows x 5 columns]

In [7]:
print(len(data[data.duplicated()])/len(data))

0.0306


</font><font color='Black'><b>Пояснение</b></font><br>
<font color='Black'>Было найдено 153 дубликата, что составляет около 3% от общего объема данных. Это допустимая величина для удаления, поэтому оставляет только оригинальные строки

In [8]:
data = data.drop_duplicates()

In [9]:
# Проверка
data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 4847 entries, 0 to 4999
Data columns (total 5 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Пол                4847 non-null   int64  
 1   Возраст            4847 non-null   float64
 2   Зарплата           4847 non-null   float64
 3   Члены семьи        4847 non-null   int64  
 4   Страховые выплаты  4847 non-null   int64  
dtypes: float64(2), int64(3)
memory usage: 227.2 KB


In [10]:
#plt.figure(figsize = (15,10))
#sns.heatmap(data.corr(), cmap="YlGnBu", annot=True)

</font><font color='Black'><b>Пояснение</b></font><br>
<font color='Black'>Теперь необходим обозначить признаки и целевой параметр. Признаки - все, что не конечный выхлоп, целевой признак - столбец со страховыми выплатами (из условия задания)

In [11]:
features = data.drop(['Страховые выплаты'], axis=1)
target = data['Страховые выплаты']

</font><font color='Black'><b>Промежуточный вывод</b></font><br>
<font color='Black'>Итак, с данными ознакомились, они представляют из себя датафрейм размером 5000 строк на 5 столбцов. В целом, при беглом первичном осмотре данных выглядят они корректно, однако было найдено 153 дубликата, от которых мы избавились. Корреляции особой не видим кроме возраста и страховых выплат (коэффициент Пирсона - 0,65). Признаки и целевой признак обозначены. Можно двигаться далее.

## Умножение матриц

В этом задании вы можете записывать формулы в *Jupyter Notebook.*

Чтобы записать формулу внутри текста, окружите её символами доллара \\$; если снаружи —  двойными символами \\$\\$. Эти формулы записываются на языке вёрстки *LaTeX.* 

Для примера мы записали формулы линейной регрессии. Можете их скопировать и отредактировать, чтобы решить задачу.

Работать в *LaTeX* необязательно.

Обозначения:

- $X$ — матрица признаков (нулевой столбец состоит из единиц)

- $y$ — вектор целевого признака

- $P$ — матрица, на которую умножаются признаки

- $w$ — вектор весов линейной регрессии (нулевой элемент равен сдвигу)

Предсказания:

$$
a = Xw
$$

Задача обучения:

$$
w = \arg\min_w MSE(Xw, y)
$$

Формула обучения:

$$
w = (X^T X)^{-1} X^T y
$$

**Ответ:** качество линейной регрессии не изменится.

**Обоснование:** итак, у нас есть предсказания - $a$. Если качество линейной регрессии не изменяется, то $a = a'$. Получается, что это и есть тезис, который здесь надо подтвердить или опровергуть.

Будем использовать следующие свойства:

$(AB)^{T}=B^{T}A^{T}$

$(AB)^{-1}=B^{-1}A^{-1}$

$AA^{-1}=A^{-1}A=E$

$AE=EA=A$

Формулы:

$a = Xw$

$a' = X'w'$, где $X'=XP$, т.е.: $a' = (XP)w'$

Исходя из вышеизложенного, чтоб соблюдалось условие $a=a'$, необходимо доказать, что $w' = P^{-1}w$.

$w = (X^T X)^{-1} X^T y$

$w' = ((XP)^{T} XP)^{-1} (XP)^{T} y=((P^{T}(X^{T}X))P)^{-1}P^{T}X^{T}y=P^{-1}(P^{T}(X^{T}X))^{-1}P^{T}X^{T}y=P^{-1}(X^{T}X)^{-1}(P^{T})^{-1}P^{T}X^{T}y=P^{-1}(X^{T}X)^{-1}EX^{T}y=P^{-1}(X^{T}X)^{-1}X^{T}y=P^{-1}w$

Что и требовалось доказать



## Алгоритм преобразования

**Алгоритм**

У нас обозначены признаки и целевой признак. Далее мы создадим матрицу размером 4х4 (сколько признаков) случайным образом и проверим ее обратимость. Обучим модель линейной регрессии и посчитаем метрику r2, которая является качественной характеристикой модели. Затем перемножим матрицу с признаками и обратимую матрицу, которые мы создали рандомным образом, так получим преобразованные данные, после чего обучим на них модель и посчитаем снова r2. В конце сравним полученные метрики r2 до и после преобразования, по идее отличий быть не должно.

**Обоснование**

Поскольку эксперимент мы будем проводить полностью согласно тому, что доказали выше, то обоснование адекватности выбранного для решения данной задачи алгоритма с теоретической точки зрения описывается в блоке 2 "Умножение матриц".

## Проверка алгоритма

In [12]:
# поскольку признака у нас четыре из условия, то и матрица должна быть такого размера
matrix = np.random.rand(4, 4)
while np.linalg.det(matrix) == 0:
    matrix = np.random.rand((4,4))
matrix

array([[0.89504295, 0.90446108, 0.64103634, 0.07872668],
       [0.95434154, 0.85479928, 0.32427208, 0.1393008 ],
       [0.4659085 , 0.78083149, 0.61952422, 0.43748953],
       [0.0912252 , 0.75168726, 0.68929667, 0.30277211]])

</font><font color='Black'><b>Пояснение</b></font><br>
<font color='Black'>Посчитаем определитель полученной матрицы

In [13]:
print(np.linalg.det(matrix))

-0.03845181684291089


</font><font color='Black'><b>Промежуточный вывод</b></font><br>
<font color='Black'>Определитель данной квадратной матрицы не равен нулю, следовательно, у данной матрицы есть обратная матрица

In [14]:
model = LinearRegression().fit(features, target)
predictions = model.predict(features)
r2 = r2_score(target, predictions)
r2

0.4302010044852066

In [15]:
matrix_2 = features @ matrix
model.fit(matrix_2, target)
predictions_2 = model.predict(matrix_2)
r2_2 = r2_score(target, predictions_2)
r2_2

0.4302010044851867

In [16]:
r2 - r2_2

1.9872992140790302e-14

</font><font color='Black'><b>Промежуточный вывод</b></font><br>
<font color='Black'>Изменение качества мизерное, его практически нет

## Финальный вывод

Задачей данного проекта была проверка факта того, изменится ли качество линейной регрессии после преобразования данных. Метрикой подсчета качества является r2. Ответ - качество линейной регрессии не изменилось, что было доказано как теоретически, так и проверено на практике.

In [17]:
matrix_checking = matrix_2 @ np.linalg.inv(matrix)
matrix_checking

0     1        2             3
0     1.000000e+00  41.0  49600.0  1.000000e+00
1    -2.786604e-12  46.0  38000.0  1.000000e+00
2    -3.782521e-13  29.0  21000.0 -2.804046e-12
3    -1.989552e-11  21.0  41700.0  2.000000e+00
4     1.000000e+00  28.0  26100.0 -4.870100e-12
...            ...   ...      ...           ...
4995 -8.891919e-12  28.0  35700.0  2.000000e+00
4996 -2.146911e-11  34.0  52400.0  1.000000e+00
4997 -1.046958e-11  20.0  33900.0  2.000000e+00
4998  1.000000e+00  22.0  32700.0  3.000000e+00
4999  1.000000e+00  28.0  40600.0  1.000000e+00

[4847 rows x 4 columns]

In [18]:
matrix_checking.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 4847 entries, 0 to 4999
Data columns (total 4 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   0       4847 non-null   float64
 1   1       4847 non-null   float64
 2   2       4847 non-null   float64
 3   3       4847 non-null   float64
dtypes: float64(4)
memory usage: 189.3 KB


In [19]:
features.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 4847 entries, 0 to 4999
Data columns (total 4 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   Пол          4847 non-null   int64  
 1   Возраст      4847 non-null   float64
 2   Зарплата     4847 non-null   float64
 3   Члены семьи  4847 non-null   int64  
dtypes: float64(2), int64(2)
memory usage: 189.3 KB


In [20]:
# возвращаение исходных названий столбцов 
matrix_checking.columns = ['Пол', 'Возраст', 'Зарплата', 'Члены семьи']

In [21]:
# избавляемся от погрешностей
features = abs(features.round(1))
matrix_checking = abs(matrix_checking.round(1))

In [22]:
# приведение к единому типу данных
features = features.astype(int)
matrix_checking = matrix_checking.astype(int)

In [23]:
# на всякий случай делаем переиндексацию (в ходе удаления дубликатов не делали)
features.reset_index(drop= True, inplace=True)
matrix_checking.reset_index(drop=True, inplace=True)

In [24]:
matrix_checking

Пол  Возраст  Зарплата  Члены семьи
0       1       41     49600            1
1       0       46     38000            1
2       0       29     21000            0
3       0       21     41700            2
4       1       28     26100            0
...   ...      ...       ...          ...
4842    0       28     35700            2
4843    0       34     52400            1
4844    0       20     33900            2
4845    1       22     32700            3
4846    1       28     40600            1

[4847 rows x 4 columns]

In [25]:
features

Пол  Возраст  Зарплата  Члены семьи
0       1       41     49600            1
1       0       46     38000            1
2       0       29     21000            0
3       0       21     41700            2
4       1       28     26100            0
...   ...      ...       ...          ...
4842    0       28     35700            2
4843    0       34     52400            1
4844    0       20     33900            2
4845    1       22     32700            3
4846    1       28     40600            1

[4847 rows x 4 columns]

In [26]:
features.equals(matrix_checking)

True

## Чек-лист проверки

Поставьте 'x' в выполненных пунктах. Далее нажмите Shift+Enter.

- [x]  Jupyter Notebook открыт
- [x]  Весь код выполняется без ошибок
- [x]  Ячейки с кодом расположены в порядке исполнения
- [x]  Выполнен шаг 1: данные загружены
- [x]  Выполнен шаг 2: получен ответ на вопрос об умножении матриц
    - [x]  Указан правильный вариант ответа
    - [x]  Вариант обоснован
- [x]  Выполнен шаг 3: предложен алгоритм преобразования
    - [x]  Алгоритм описан
    - [x]  Алгоритм обоснован
- [x]  Выполнен шаг 4: алгоритм проверен
    - [x]  Алгоритм реализован
    - [x]  Проведено сравнение качества моделей до и после преобразования